socorro/signature/__main__.py

```diff
@@ -0,0 +1,260 @@
+    parser.add_argument(
+        '-v', '--verbose', help='increase output verbosity', action='store_true'
+    )
+    parser.add_argument(
+        '--format', help='specify output format: csv, text (default)'
+    )
+    parser.add_argument(
+        'crashids', metavar='crashid', nargs='+', help='crash id to generate signatures for'
+    )
+    args = parser.parse_args()
+
+    if args.format == 'csv':
+        outputter = CSVOutput
+    else:
+        outputter = TextOutput
+
+    if args.verbose:
+        logging_level = 'DEBUG'
+    else:
+        logging_level = 'INFO'
+
+    api_token = os.environ.get('SOCORRO_API_TOKEN', '')
+
+    setup_logging(logging_level)
+
+    generator = SignatureGenerator(debug=args.verbose)
+
+    with outputter() as out:
+        for crash_id in args.crashids:
+            resp = fetch('/RawCrash/', crash_id, api_token)
+            if resp.status_code == 404:
+                out.warning('%s: does not exist.' % crash_id)
+                continue
+            if resp.status_code == 429:
+                out.warning('API rate limit reached. %s' % resp.content)
+                # FIXME(willkg): Maybe there's something better we could do here. Like maybe wait a
+                # few minutes.
+                return 1
+            if resp.status_code == 500:
+                out.warning('HTTP 500: %s' % resp.content)
+                continue
+
+            raw_crash = resp.json()
+
+            # If there's an error in the raw crash, then something is wrong--probably with the API
+            # token. So print that out and exit.
+            if 'error' in raw_crash:
+                out.warning('Error fetching raw crash: %s' % raw_crash['error'])
+                return 1
+
+            raw_crash_minimal = {
+                'JavaStackTrace': raw_crash.get('JavaStackTrace', None),
+                'OOMAllocationSize': raw_crash.get('OOMAllocationSize', None),
+                'AbortMessage': raw_crash.get('AbortMessage', None),
+                'AsyncShutdownTimeout': raw_crash.get('AsyncShutdownTimeout', None),
+                'ipc_channel_error': raw_crash.get('ipc_channel_error', None),
+                'additional_minidumps': raw_crash.get('additional_minidumps', None),
+                'IPCMessageName': raw_crash.get('IPCMessageName', None),
+            }
+
+            resp = fetch('/ProcessedCrash/', crash_id, api_token)
+            if resp.status_code == 404:
+                out.warning('%s: does not have processed crash.' % crash_id)
+                continue
+            if resp.status_code == 429:
+                out.warning('API rate limit reached. %s' % resp.content)
+                # FIXME(willkg): Maybe there's something better we could do here. Like maybe wait a
+                # few minutes.
+                return 1
+            if resp.status_code == 500:
+                out.warning('HTTP 500: %s' % resp.content)
+                continue
```

Comment:
There's a lot of code repetition now. And perhaps there should be a catch all kinda thing for all other non-200.

Reply:
It's six lines of code repeated. In situations like this where the repeated code is small and there's only one repeat, I don't bother with extracting it out and then dealing with that. The delta would be a couple of saved lines, but it introduces unneeded maintenance burden and abstractions and all that.

What other non-200 response codes would we get back from AWS S3 via HTTP?

Reply:
> What other non-200 response codes would we get back from AWS S3 via HTTP?

AWS S3?? This is from the webapp. 

No other error is "expected". If the webapp is busted at the moment (but ELB and Nginx working) you might get a 502 but if that's the case all bets are off.

Reply:
Oh. I figured you would know all the responses the webapp did, so for ones you didn't know about maybe API endpoints that pull from S3, just passed the HTTP response along.